feat: add home view section screen

MounirDhahri opened the pull request:
**[Ready for view as of Monday 9th, September]**
 
This PR resolves [ONYX-1159]


### Description
This PR adds a new screen generic type that will render a list of artworks. 


https://github.com/user-attachments/assets/a824fe2f-0304-428b-8b05-26557d58b197



https://github.com/user-attachments/assets/6aad0a6a-85f8-4070-b903-b01703f9af58





<!-- Info, implementation, how to get there, before & after screenshots & videos, follow-up work, etc -->

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
<!-- ⓘ 'User-facing' changes will be published as release notes. -->
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->
<!-- ⚠️ Prefix with `[NEEDS EXTERNAL QA]` if a change requires external QA -->

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

<!-- end_changelog_updates -->

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md


[ONYX-1159]: https://artsyproduct.atlassian.net/browse/ONYX-1159?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

MounirDhahri commented on src/app/Scenes/HomeSection/HomeSection.tsx:
```diff
@@ -0,0 +1,64 @@
+}
+
+export const HomeSectionScreen: React.FC<HomeSectionScreenProps> = ({ section }) => {
+  const title =
```
there is a probably a more beautiful way to get this

MounirDhahri commented on src/app/Scenes/HomeSection/HomeSection.tsx:
```diff
@@ -0,0 +1,64 @@
+
+      <Screen.Body fullwidth>
+        {section?.__typename === "ArtworksRailHomeViewSection" && (
+          <ArtworksScreenHomeSectionQR sectionId={section.internalID} />
```
This is a nested `QueryRenderer` - not because I am trying to optimise loading from the first go but it's because I copied the logic I made for the old NWFY.

MounirDhahri commented on src/app/routes.ts:
```diff
@@ -192,6 +192,7 @@ function getDomainMap(): Record<string, RouteMatcher[] | null> {
     addRoute("/galleries-for-you", "GalleriesForYou"),
     addRoute("/gene/:geneID", "Gene"),
     addRoute("/home-view", "HomeView"),
+    addRoute("/home-view/section/:sectionId", "HomeSectionScreen"),
```
glad to hear thoughts about the naming here

dblandin replied:
**suggestion**: `sectionId` > `sectionID`

MounirDhahri replied:
nice idea

dblandin commented on src/app/Scenes/HomeSection/ArtworksScreenHomeSection.tsx:
```diff
@@ -0,0 +1,130 @@
+    return (
+      <ArtworksScreenHomeSection
+        section={data.homeView.section}
+        title={data.homeView.section.component.title}
```
**question**: Since we're passing the full section above, do we need a separate `title` prop that relies on nested data?

MounirDhahri replied:
good point

olerichter00 commented on src/app/Scenes/HomeViewSectionScreen/Artworks/HomeViewSectionScreenArtworks.tsx:
```diff
@@ -0,0 +1,129 @@
+      artworksQuery,
+      { id: props.sectionId },
+      {
+        // This is necessary to avoid displaying cached because this component is used in more than one place and Relay returns data from another screen before the new data is loaded.
```
Shouldn't this return cached data per sectionID? 🤔

olerichter00 commented on src/app/Scenes/HomeViewSectionScreen/Artworks/HomeViewSectionScreenArtworks.tsx:
```diff
@@ -0,0 +1,129 @@
+
+    // This won't happen because the query would fail thanks to the @principalField
+    // Adding it here to make TS happy
+    if (!data.homeView.section || !data.homeView.section.component?.title) {
```
If we require the title here, should we make it nonNullable in the schema instead?

dblandin replied:
I believe `title` was already non-nullable but [this PR](https://github.com/artsy/metaphysics/pull/5881) should alleviate the other null checking here.

dblandin commented on src/app/Scenes/HomeViewSectionScreen/Artworks/HomeViewSectionScreenArtworks.tsx:
```diff
@@ -0,0 +1,129 @@
+import { pluralize } from "app/utils/pluralize"
+import { graphql, useLazyLoadQuery, usePaginationFragment } from "react-relay"
+
+export const PAGE_SIZE = 100
```
**question**: What would unlock using a smaller initial page size and relying on pagination triggers as a user scrolls?

dblandin replied:
I just merged https://github.com/artsy/eigen/pull/10555 which partially addresses this comment. Let's make sure that we have similar pagination for the new screen as well.

dblandin reviewed (commented):
Looking good! 🔥

dblandin commented on src/app/Scenes/HomeView/Sections/ArtworksRailHomeViewSection.tsx:
```diff
@@ -17,7 +17,7 @@ export const ArtworksRailHomeViewSection: React.FC<ArtworksRailHomeViewSectionPr
   const title = data.component?.title
   const artworks = extractNodes(data.artworksConnection)
-  const componentHref = "" // TODO: should be in schema
+  const componentHref = `/home-view/section/${data.internalID}`
```
**suggestion**: This could be inlined within the call to `navigate()` below until we model this (or something like behaviors > viewAll) in the schema.

MounirDhahri replied:
It's used in two contexts (line 20 and line 44), that's why I kept it exported as a variable

dblandin commented on src/app/Scenes/HomeViewSectionScreen/Artworks/HomeViewSectionScreenArtworks.tsx:
```diff
@@ -0,0 +1,129 @@
+}
+
+export const ArtworksScreenHomeSection: React.FC<ArtworksScreenHomeSection> = ({ section }) => {
+  const defaultViewOption = GlobalStore.useAppState((state) => state.userPrefs.defaultViewOption)
```
**question**: Without having a control to toggle between the two options, should we explicitly pick a view style rather than referring to this default?

dblandin commented on src/app/Scenes/HomeViewSectionScreen/Artworks/HomeViewSectionScreenArtworks.tsx:
```diff
@@ -0,0 +1,129 @@
+        pageSize={PAGE_SIZE}
+        contextModule={ContextModule.newWorksForYouRail}
+        contextScreenOwnerType={OwnerType.newWorksForYou}
+        contextScreen={OwnerType.newWorksForYou}
```
**suggestion**: We should figure out the schema support for passing these values for analytics. In the meantime, you could choose something generic related to the section type if the underlying component requires these props.

MounirDhahri commented on src/app/Scenes/HomeViewSectionScreen/Artworks/HomeViewSectionScreenArtworks.tsx:
```diff
@@ -0,0 +1,129 @@
+        // This is necessary to avoid displaying cached because this component is used in more than one place and Relay returns data from another screen before the new data is loaded.
+        fetchPolicy: "network-only",
+      }
+    )
```
Agree- maybe this is something to look into- tbh it wasn't an issue when I built this. I mainly copied the code here but probably it's not needed anymore https://github.com/artsy/eigen/blob/70571f97c97fb86cd8f0e997496107e33602f693/src/app/Components/WorksForYouArtworks.tsx#L126

olerichter00 commented on src/app/Scenes/HomeViewSectionScreen/Artworks/HomeViewSectionScreenArtworks.tsx:
```diff
@@ -0,0 +1,115 @@
+  section: HomeViewSectionScreenArtworks_artworksRailHomeViewSection$key
+}
+
+export const ArtworksScreenHomeSection: React.FC<ArtworksScreenHomeSection> = (props) => {
```
**Suggestion:** Maybe we can give the file the same name as the component that it exports to make it easier to find.

anandaroop commented on src/app/Scenes/HomeView/Sections/ArtworksRailHomeViewSection.tsx:
```diff
@@ -22,7 +22,7 @@ export const ArtworksRailHomeViewSection: React.FC<ArtworksRailHomeViewSectionPr
   const title = data.component?.title
   const artworks = extractNodes(data.artworksConnection)
-  const componentHref = data.component?.behaviors?.viewAll?.href
+  const componentHref = "home-view/sections/" + data.internalID
```
**question:** Shouldn't this prefer to navigate to the `viewAll.href` **_if_** it is provided, and only provide the generic full-screen experience here if `viewAll.href` is falsy?

nickskalkin replied:
Yes, I think we were agreeing on this behavior (response 2): https://github.com/artsy/metaphysics/pull/5963#issuecomment-2326108590

MounirDhahri replied:
Good point - I will the MP changes to reflect that the default behaviour is what we need here.

MounirDhahri replied:
https://github.com/artsy/metaphysics/pull/5988

MounirDhahri replied:
I made some changes [here](https://github.com/artsy/eigen/pull/10546/commits/7fd8756a50a36e41d15785a997dc9d9d6748e8c5) and added a new helper for future use 
<img width="510" alt="Screenshot 2024-09-10 at 11 27 23" src="https://github.com/user-attachments/assets/adf306ab-3ef1-4856-ae01-aea0e721adef">

anandaroop reviewed (commented):
Looks good! Just one question below

MounirDhahri commented on src/app/routes.ts:
```diff
@@ -192,6 +192,7 @@ function getDomainMap(): Record<string, RouteMatcher[] | null> {
     addRoute("/galleries-for-you", "GalleriesForYou"),
     addRoute("/gene/:geneID", "Gene"),
     addRoute("/home-view", "HomeView"),
+    addRoute("/home-view/sections/:sectionID", "HomeViewSectionScreen"),
```
Yes, that's in the pipeline for next quarter with Sapphire. We are thinking of adjusting this file and making things easier here

MounirDhahri commented:
Going to merge this PR before I need to merge conflicts from 10 commits - I addressed all review comments above and will be glad to address any others that come up
